### PR TITLE
TandemFoil: gc=0.2 sweep (continuing softening trend)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TF gc sweep has shown monotonic improvement: gc=1.0→0.5→0.3 all improve. The new champion (PR #3108, zenitsu) uses gc=0.3 and achieved val=21.909 (-2.8%). Testing gc=0.2 (and gc=0.1 as a limit check) continues the trend.

At some point gc becomes too soft and the model destabilizes (similar to TFP where gc=0.3 was too tight but gc=0.5 was correct). The question is where that floor is for TF.

## Instructions

**Run 1: gc=0.2**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.2 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group tf-gc-sweep
```

**Run 2: gc=0.1 (limit check)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.1 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group tf-gc-sweep
```

## Baseline (UPDATED — PR #3108)
- **val_primary/surface_pressure_mae:** 21.909 at epoch 334 (gc=0.3)
- **test_primary/surface_pressure_mae:** 23.419

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT usopp:

## Results

### Run 1: gc=0.2 — NEW BEST

| Metric | Baseline (gc=0.3) | gc=0.2 | Delta |
|---|---|---|---|
| **val_primary/surface_pressure_mae** | 21.909 (ep334) | **21.350** (ep331) | **-2.6%** |
| **test_primary/surface_pressure_mae** | 23.419 | **23.195** | **-1.0%** |
| test geom_camber_cruise | 27.436 | 27.211 | -0.8% |
| test geom_camber_rc | 31.587 | 32.297 | +2.2% |
| test re_rand | 17.119 | 16.824 | -1.7% |
| test single_in_dist | 17.533 | 16.448 | -6.2% |

- **W&B run:** [9g11l7tm](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/9g11l7tm)
- **Peak memory:** ~4.6 GB (GPU 0)
- **Epochs:** 341 (360-min timeout), best at ep331
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.2 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --agent usopp --wandb-name "usopp/tf-gc02" --wandb-group tf-gc-sweep
```

### Run 2: gc=0.1 (limit check) — WORSE

| Metric | Baseline (gc=0.3) | gc=0.1 | Delta |
|---|---|---|---|
| **val_primary/surface_pressure_mae** | 21.909 (ep334) | **22.141** (ep335) | **+1.1%** |
| **test_primary/surface_pressure_mae** | 23.419 | **23.797** | **+1.6%** |
| test geom_camber_cruise | 27.436 | 26.751 | -2.5% |
| test geom_camber_rc | 31.587 | 31.861 | +0.9% |
| test re_rand | 17.119 | 16.861 | -1.5% |
| test single_in_dist | 17.533 | 19.716 | +12.4% |

- **W&B run:** [55xdv4fq](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/55xdv4fq)
- **Peak memory:** ~4.6 GB (GPU 1)
- **Epochs:** 341 (360-min timeout), best at ep335
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.1 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --agent usopp --wandb-name "usopp/tf-gc01" --wandb-group tf-gc-sweep
```

## What happened

The gc softening trend continues: **gc=0.2 sets a new best** with val=21.350 (-2.6% vs gc=0.3 baseline). The improvement is consistent across most splits, with single_in_dist showing the strongest gain (-6.2%).

**gc=0.1 confirms the floor.** At gc=0.1, the model regresses to 22.141 (+1.1%). The key failure mode is on single_in_dist (+12.4%), suggesting that extremely soft clipping allows the model to overfit to distribution-specific gradients. Interestingly, gc=0.1 still beats gc=0.3 on some splits (geom_camber_cruise, re_rand), but the aggregate metric is worse.

The gc sweep now shows a clear curve: gc=1.0 → 0.5 → 0.3 → **0.2** (all improve monotonically), then 0.1 regresses. **gc=0.2 is the optimal clip value for TandemFoil under EMA=0.999 + Lion lr=1.25e-4.**

Both runs trained ~341 epochs in 360 minutes, similar to the gc=0.3 baseline (334 epochs), so throughput is equivalent — the improvement is purely from the optimization dynamics.

## Suggested follow-ups

1. **gc=0.15 or gc=0.25 fine-grained sweep** — the optimal may be between 0.2 and 0.3, a finer grid could squeeze out another fraction.
2. **gc=0.2 + longer T_max** — the T_max=10 cycling combined with very soft gc may benefit from a slightly longer cycle (T_max=15 or 20).
3. **Transfer gc=0.2 to TandemFoil Paper** — current TFP champion uses gc=0.5; gc=0.2 may transfer the improvement.

**Comment by tcapelle** [comment]
ADVISOR: Merging — gc=0.2 is the new TF champion. 21.350 (-2.6% vs 21.909). gc trend confirmed monotonic: 1.0→0.5→0.3→0.2. gc=0.1 overshoots (22.141, single_in_dist regression +12.4%), so 0.2 is the floor. Test metric 23.195 (-1.0%). Model not yet converged — further improvements available.
